### PR TITLE
Remove IAM from db

### PR DESCRIPTION
**Pull request by Jeasmine**
* Remove IAMs with redisplay from DB when lastDisplay is
    older than six month

---
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/951)

### CODE REVIEW

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java [hunk 1]
```diff
@@ -154,16 +155,26 @@ void receivedInAppMessageJson(@NonNull JSONArray json) throws JSONException {
         // Cache copy for quick cold starts
         OneSignalPrefs.saveString(OneSignalPrefs.PREFS_ONESIGNAL,
            OneSignalPrefs.PREFS_OS_CACHED_IAMS, json.toString());
-        processInAppMessageJson(json);
+        ArrayList<OSInAppMessage> newMessages = processInAppMessageJson(json);
+
+        List<OSInAppMessage> updatedRedisplayMessages = inAppMessageRepository.updateInAppMessage(newMessages, redisplayInAppMessages);
```
Comment (jkasten2) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java]: The source of `newMessages` is driven directly by the results from the  `on_session` network call . However the results from `on_session` is drive by the backend and if the player fits into the Segments defined there. So this means the same IAM could fall in and out of Segment multiple times which would result in the on_session not always have it either.

So we don't reset the display stats noted above we should only delete these local records after a long period of time to be safe or not at all. Can we set this to 6 months to be safe?

Comment (Jeasmine) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java]: oh ok, i will address this on here and iOS. 6 Months seems like a logic time period

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java [hunk 2]
```diff
@@ -154,16 +155,26 @@ void receivedInAppMessageJson(@NonNull JSONArray json) throws JSONException {
...
         ArrayList<OSInAppMessage> newMessages = new ArrayList<>();
         for (int i = 0; i < json.length(); i++) {
             JSONObject messageJson = json.getJSONObject(i);
             OSInAppMessage message = new OSInAppMessage(messageJson);
             newMessages.add(message);
         }
+        return newMessages;
+    }
+
+    private void setMessages( ArrayList<OSInAppMessage> newMessages) {
```
Comment (mikechoch) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java]: Remove space before `ArrayList`

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java [hunk 3]
```diff
@@ -472,14 +483,14 @@ public void run() {
 
         //Update the data to enable future re displays
```
Comment (mikechoch) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java]: Add a space after `//`

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java [hunk 4]
```diff
@@ -472,14 +483,14 @@ public void run() {
...
         //Avoid calling the repository data again
```
Comment (mikechoch) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java]: Add a space after `//`

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java [hunk 5]
```diff
@@ -572,7 +583,7 @@ public void messageTriggerConditionChanged() {
      * */
```
Comment (mikechoch) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java]: `* */` should just be `*/`

File: OneSignalSDK/unittest/src/test/java/com/test/onesignal/TestHelpers.java [hunk 6]
```diff
@@ -373,6 +398,49 @@ else if (type == Cursor.FIELD_TYPE_FLOAT)
       return iams;
    }
 
+   @WorkerThread
+   static void deleteOldInAppMessages() {
```
Comment (jkasten2) [inline OneSignalSDK/unittest/src/test/java/com/test/onesignal/TestHelpers.java]: This is duplicated, it exists in `com.onesignal.OSInAppMessageRepository` as well. Also I don't see this being called in the tests so I think we can remove this `deleteOldInAppMessages` method here.

File: OneSignalSDK/unittest/src/test/java/com/test/onesignal/TestHelpers.java [hunk 7]
```diff
@@ -373,6 +398,49 @@ else if (type == Cursor.FIELD_TYPE_FLOAT)
...
+            String idToRemove = cursor.getString(cursor.getColumnIndex(OneSignalPackagePrivateHelper.InAppMessageTable.COLUMN_NAME_MESSAGE_ID));
+            deleteInAppMessage(idToRemove);
+         } while (cursor.moveToNext());
+      }
+
+      if (!cursor.isClosed())
+         cursor.close();
+   }
+
+   static void deleteInAppMessage(String messageId) {
```
Comment (jkasten2) [inline OneSignalSDK/unittest/src/test/java/com/test/onesignal/TestHelpers.java]: This is duplicated, it exists in `com.onesignal.OSInAppMessageRepository` as well. Also I don't see this being called in the tests so I think we can remove this `deleteInAppMessage` method here.

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageRepository.java [hunk 8]
```diff
@@ -15,12 +16,40 @@
...
     OSInAppMessageRepository(OneSignalDbHelper dbHelper) {
         this.dbHelper = dbHelper;
     }
 
+    /**
+     * Remove IAMs that the last display time was six month ago
+     */
+    @WorkerThread
+    void deleteOldInAppMessages(long dateInSeconds) {
+        SQLiteDatabase readableDb = dbHelper.getReadableDbWithRetries();
```
Comment (jkasten2) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageRepository.java]: We could use a writable db here and performance a single write. Instead getting a list of ids then doing a delete on each record one at a time.

File: OneSignalSDK/unittest/src/test/java/com/test/onesignal/TestHelpers.java (hunk 6 above)
Comment (Jeasmine) [reply OneSignalSDK/unittest/src/test/java/com/test/onesignal/TestHelpers.java]: good catch, this was used before but I changed it. TY!

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageRepository.java (hunk 8 above)
Comment (Jeasmine) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageRepository.java]: Absolutely right

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java [hunk 9]
```diff
@@ -144,6 +145,15 @@ void receivedInAppMessageJson(@NonNull JSONArray json) throws JSONException {
...
         processInAppMessageJson(json);
+
+        // Remove cache redisplayed messages
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                Thread.currentThread().setPriority(Process.THREAD_PRIORITY_BACKGROUND);
+                inAppMessageRepository.deleteOldRedisplayedInAppMessages();
+            }
+        }, OS_SAVE_IN_APP_MESSAGE).start();
```
Comment (mikechoch) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java]: Change the tag associated with the thread, this is not a save action is it?

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageDisplayStats.java [hunk 10]
```diff
@@ -87,13 +87,10 @@ boolean shouldDisplayAgain() {
...
     boolean isDelayTimeSatisfied() {
-        long currentTimeInSeconds = System.currentTimeMillis() / 1000;
-
         if (lastDisplayTime < 0) {
-            lastDisplayTime = currentTimeInSeconds;
             return true;
         }
-
+        long currentTimeInSeconds = System.currentTimeMillis() / 1000;
         //Calculate gap between display times
```
Comment (mikechoch) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageDisplayStats.java]: Space after `//`

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageRepository.java [hunk 11]
```diff
@@ -15,32 +15,24 @@
...
-            if (writableDb != null) {
-                try {
-                    writableDb.endTransaction(); // May throw if transaction was never opened or DB is full.
-                } catch (Throwable t) {
-                    OneSignal.Log(OneSignal.LOG_LEVEL.ERROR, "Error closing transaction! ", t);
-                }
-            }
-        }
+        writableDb.delete(OneSignalDbContract.InAppMessageTable.TABLE_NAME,
+                OneSignalDbContract.InAppMessageTable.COLUMN_NAME_LAST_DISPLAY + "<?",
```
Comment (mikechoch) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageRepository.java]: Space between `<?` -> `< ?`

**Pull request merged**